src/librustc_resolve/lib.rs

```diff
@@ -3457,6 +3521,98 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         found_traits
     }
 
+    // Looks up a definitions (and paths) based on a user-supplied predicate
+    // This method should be used to generate suggestions when name resolution
+    // fails.
+    fn lookup_suggestions<FilterFn>(&mut self,
+                                    lookup_name: Name,
+                                    lookup_type: LookupTarget,
+                                    namespace: Namespace,
+                                    filter_fn: FilterFn) -> Vec<Path>
+        where FilterFn: Fn(Def) -> bool {
+
+        let mut lookup_results: Vec<Path> = Vec::new();
+        let mut lookup_modules: Vec<(Module, Vec<PathSegment>)> = Vec::new();
+        lookup_modules.push((self.graph_root, Vec::new()));
+
+        while let Some((in_module, path_segments)) = lookup_modules.pop() {
+            build_reduced_graph::populate_module_if_necessary(self, &in_module);
+
+            in_module.for_each_child(|_, ns, name_binding| {
+
+                // collect results based on the filter function
+                if ns == namespace {
+                    if let Some(def) = name_binding.def() {
+                        if let Some(def_id) = def.def_id_opt() {
+                            if filter_fn(def) {
+                                // find the name
+                                let name = self.get_defid_name(def_id);
```

Comment:
This is wrong if `name_binding` was imported under a different name.
Also, `for_each_child` already provides the correct name as the first argument to this closure, making `get_defid_name` unnecessary.